Mage.Sets/src/mage/cards/a/AminatousAugury.java

```diff
@@ -134,54 +166,151 @@ public AminatousAuguryCastFromExileEffect copy() {
-                    }
+
+                // make all cards with eligible types zero mana cost
+                if (!Collections.disjoint(unusedCardTypes,cardToCheck.getCardType())) {
+                    AminatousAuguryCost aminatouCost = new AminatousAuguryCost();
+                    Costs costs = new CostsImpl();
+                    costs.add(aminatouCost);
+                    costs.addAll(affectedAbility.getCosts());
+                    player.setCastSourceIdWithAlternateMana(affectedAbility.getSourceId(), null, costs);
                     return true;
                 }
             }
         }
         return false;
     }
 }
+
+
+class AminatousAuguryCost extends CostImpl {
+
+    AminatousAuguryCost() {
+        this.text = "for each nonland cardtype, you may cast a card of that type from among the exiled cards without paying its mana cost";
+    }
+
+    AminatousAuguryCost(final AminatousAuguryCost cost) {
+        super(cost);
+    }
+
+    @Override
+    public boolean pay(Ability ability, Game game, UUID sourceId, UUID controllerId, boolean noMana, Cost costToPay) {
+
+        /*
+         Use map to match card to the Aminatou card used to exile it. Then use that Aminatou card Id,
+         to load castable card types
+        */
+        HashMap<UUID,UUID> setLocater = (HashMap<UUID, UUID>) game.getState().getValue("cardToAminatouIdMap");
+        String aminatouId = setLocater.get(sourceId).toString() + "_unusedCardTypes";
+        EnumSet<CardType> unusedCardTypes = (EnumSet<CardType>) game.getState().getValue(aminatouId);
+        if (unusedCardTypes == null){
+            throw new UnsupportedOperationException("Could not load unused card types for Aminatou's Augury");
+        }
+
+        Set<CardType> castCardTypes = ability.getSourceObject(game).getCardType();
+        if (castCardTypes != null) {
+
+            // Count matching available card types
+            CardType useType = null;
+            int numAvailTypes = 0;
+            for (CardType chkType : castCardTypes){
+                if (unusedCardTypes.contains(chkType)){
+                    numAvailTypes++;
+                    useType = chkType;
+                }
+            }
+
+            if (numAvailTypes == 0){
+                // If no available types, then do not cast.
+                paid = false;
+            } else if (numAvailTypes == 1){
+                // If card has multiple types, and only one is unused,
+                // then cast using that type OR if card has only one type
+                // and it is available, then cast.
+                unusedCardTypes.remove(useType);
+                paid = true;
+            } else if (numAvailTypes >= 2){
+                // If card has multiple types, and more than one has not yet been used,
+                // then show dialog to select which type to use for cast.
+                AminatousAuguryCardTypeChoice choices = new AminatousAuguryCardTypeChoice(castCardTypes);
```

Comment:
You must filter unused types (compare card type and used types) and make it selectable. Current versions allow to select any type from card type even if it was used before.